Add fsync for directory

icristescu opened the pull request:
Calling `fsync` on newly created files is not enough - the corresponding entry in the containing directory might not make to disk. An explicit `fsync()` on a file descriptor for the directory is also needed (from https://man7.org/linux/man-pages/man2/fsync.2.html). 

We need this for instance after creating the snapshot directory in #2125. 

I am not very confident in this patch, so any feedback is very welcome.

icristescu commented:
Rebased over https://github.com/mirage/irmin/pull/2125 to include the usage for this function.

Ngoguey42 commented:
I now understand what you were trying to achieve. I covers the situation where the computer would power out just after the end of the snapshot export program.

I'm not against that diff.

Another solution would maybe be to put all these fsyncs in repo's close routine

metanivek commented on src/irmin-pack/unix/s.ml:
```diff
@@ -53,6 +53,18 @@ module type S = sig
+      the existing one, containing only one commit, specified by the [key]. Note
+      that this operation is blocking.
+
+      If [repo] is opened with the flag [use_fsync] set, then [path] is flushed
```
```suggestion
      If [repo] is opened with the flag [use_fsync] set, then [path] is fsynced
```

metanivek commented on src/irmin-pack/unix/ext.ml:
```diff
@@ -351,6 +354,36 @@ module Maker (Config : Conf.S) = struct
+            let* branch_store =
+              Branch.v ~fresh:true ~readonly:false branch_path
+            in
+            (* Gc worker always fsyncs prefix and mapping files. File_manager
+               fsyncs dict, control, suffix and index. Fsync branch file here. Then
+               fsync directory. *)
+            let use_fsync = Irmin_pack.Conf.use_fsync t.config in
+            let () = if use_fsync then Branch.fsync branch_store in
+            let* () = Branch.close branch_store in
+            let () = Io.fsync_dir path |> Errs.raise_if_error in
```
For a future when file manager also manages the branch store (https://github.com/mirage/irmin/pull/2125#discussion_r1015589263), all this code could live in `File_manager.create_one_commit_store` which would be cleaner and only require a comment about GC handling prefix/mapping files.

icristescu replied:
yes, agree, these sort of comment show that there is an issue in the structure of the code

metanivek commented on src/irmin-pack/unix/io.ml:
```diff
@@ -290,4 +290,14 @@ module Unix = struct
+  let fsync_dir path =
+    try
+      let dirfd =
+        Unix.openfile path Unix.[ O_RDONLY; O_SYNC; O_RSYNC ] default_open_perm
```
It probably doesn't matter but I'm not sure that we need the two `*SYNC` flags when opening a directory.

icristescu replied:
I'm not sure either :) I read the doc on `O_RSYNC` (`reads complete as writes (depending on O_SYNC/O_DSYNC)`) as in both flags are necessary.

metanivek replied:
I read a little more on these flags (like https://www.man7.org/linux/man-pages/man2/open.2.html).

I think you are right that `O_RSYNC` depends on the presence of either `O_SYNC` or `O_DSYNC`, but it appears they only apply to regular files. So I think we can remove them when opening a file descriptor for a directory.

I also learned that `O_RSYNC` isn't implemented in Linux
> Linux implements O_SYNC and O_DSYNC, but not O_RSYNC.  Somewhat
       incorrectly, glibc defines O_RSYNC to have the same value as
       O_SYNC.  (O_RSYNC is defined in the Linux header file
       <asm/fcntl.h> on HP PA-RISC, but it is not used.)

but I found this for the [IBM i operating system](https://www.ibm.com/docs/en/i/7.5?topic=ssw_ibm_i_75/apis/open.html)

> O_RSYNC
>    Read operations to the file will be performed synchronously. Pending update requests affecting the data to be read are written to permanent storage. This flag is used in combination with O_SYNC or O_DSYNC. When O_RSYNC and O_SYNC are set, all file data and file attributes are written to permanent storage before the read operation returns. When O_RSYNC and O_DSYNC are set, all file data is written to permanent storage before the read operation returns.

metanivek commented on src/irmin-pack/unix/io.ml:
```diff
@@ -290,4 +290,14 @@ module Unix = struct
+        Unix.openfile path Unix.[ O_RDONLY; O_SYNC; O_RSYNC ] default_open_perm
+      in
+      Unix.fsync dirfd;
+      Unix.close dirfd;
```
We'd also want to close the file descriptor if an exception is raised to avoid a leak.

icristescu replied:
yes, thanks!

metanivek reviewed (commented):
:+1: I'm in favor of making our fsyncs more robust. Once this lands, lets consider modifying the gc worker to do the same.

icristescu commented:
> I now understand what you were trying to achieve. I covers the situation where the computer would power out just after the end of the snapshot export program.

Hm, I though it was necessary, but reading your comment, I realise that it's maybe an overkill :) But the `fsync_dir` function is necessary in the case of `fsync` files created by `split`. I can keep only the `fsync_dir` function and remove its call from the export.

Ngoguey42 commented:
My understanding of fsync is that calling it is only necessary to ensure persistence for situations where the system brutally turns off. In other words: if the computer never shuts down, fsync can be seen as a no-op.

The page cache being shared by all processes, it ensures that the files created by a process are immediately visible by other processes. The same goes for modifications of directories: A modification to a directory is first reflected in the page cache before being persisted to disk

icristescu commented:
I removed the fsync_dir for the export snapshot, and added where I think it would make maybe more sense (in the gc_worker, where new files are created). But please feel free to close this PR if you think its unnecessary.

metanivek commented:
@Ngoguey42 you're correct in that `fsync` (and friends) is used to _attempt_ durability guarantees and/or ordering of writes by forcing data to physical storage when it is called. I say _attempt_ because it seems (based on the research I've done) to be dependent on file systems and actual storage device setups (some might not actually flush to disk, for example). 

I have no strong opinion of whether we want to have this kind of durability guarantee in a situation like the snapshot export. My hunch is maybe it does not matter.

I think that adding `fsync` of a directory when new files have been added serves one main purpose. It is a more correct `fsync` strategy when a store has been configured to use `fsync` since we know that `fsync`ing just the file in this case is not sufficient (for some file systems / mount options :sweat_smile:). Related, if we want to `fsync` files in the GC worker that have been created, we ought to also `fsync` the directory.

From [Ensuring data reaches disk](https://lwn.net/Articles/457667/):
> The more subtle usages deal with newly created files, or overwriting existing files. A newly created file may require an fsync() of not just the file itself, but also of the directory in which it was created (since this is where the file system looks to find your file). This behavior is actually file system (and mount option) dependent. You can either code specifically for each file system and mount option combination, or just perform fsync() calls on the directories to ensure that your code is portable.

Ngoguey42 commented:
Also, I believe that the design i've written there https://github.com/mirage/irmin/issues/2082 can improve (ensure?) consistency without any fsync call.
